File: 206-reverse-linked-list/206-reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    
    void rev(ListNode* temp, ListNode* &head)
    {
        if(temp->next == NULL)
            return;
        
        temp = temp->next;
        rev(temp,head);
        
        cout<<temp->val<<endl;
        
        head->val = temp->val;
        head->next = new ListNode();
        head = head->next;
    }
    
    ListNode* reverseList(ListNode* head) 
    {
        if(head == NULL)
            return NULL;
        
        ListNode* temp = head;
        ListNode* ans = new ListNode();
        ListNode* ref = ans;
        
        rev(temp, ans);
        ans->val = head->val;
        ans->next = NULL;
        
        return ref;
    }
};

File: 206-reverse-linked-list/206. Reverse Linked List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    ListNode temp;

    public ListNode recur(ListNode head) {
        if (head == null)
            return null;
        if (head.next == null)
        {
            if(temp == null)temp = head;
            IO.println(temp.val);
            return head;
        }

        ListNode parent = recur(head.next);
        parent.next = head;
        head.next = null;

        return head;
    }

    public ListNode reverseList(ListNode head) {

        recur(head);

        return temp;
    }
}